### PR TITLE
feat(ComboBox): L3-6409 Combo box Component

### PR DESCRIPTION
**Pull request by zhenpanda**
**Jira ticket**

[L3-6409](https://phillipsauctions.atlassian.net/browse/L3-6409)
[L3-5961](https://phillipsauctions.atlassian.net/browse/L3-5961)

https://github.com/user-attachments/assets/5982e59c-bfdc-4224-875b-7bc73cd3b276

https://github.com/user-attachments/assets/782daa09-508a-422f-a499-7902660fbfc1

**Summary**

Added Combo Box component
Added Phone Number Picker from Combo Box component

**Change List (describe the changes made to the files)**

This pull request introduces two new components, `ComboBox` and `PhoneNumberPicker`, along with their corresponding styles, tests, and Storybook entries. It also updates dependencies and modifies shared styles to support these components. Below is a summary of the most important changes:

### New Components
* **`ComboBox` Component**: Added a reusable `ComboBox` component with customizable options, input handling, and dropdown functionality. Includes accessibility features and integration with the `cmdk` and `@radix-ui/react-popover` libraries. (`src/components/ComboBox/ComboBox.tsx`, `[src/components/ComboBox/ComboBox.tsxR1-R173](diffhunk://#diff-e4a18ca4c1c01736ac00eacc466bec2e8d90df67f8bc5c47b58bac1e981270b8R1-R173)`)
* **`PhoneNumberPicker` Component**: Built on top of the `ComboBox`, this component allows users to select a country code for phone numbers. It leverages the `libphonenumber-js` library for country data. (`src/components/PhoneNumberPicker/PhoneNumberPicker.tsx`, `[src/components/PhoneNumberPicker/PhoneNumberPicker.tsxR1-R66](diffhunk://#diff-d33d4ed6dd0aeb79b78a806ed1995e249899a6bbfc142f6c7e5163546fb79927R1-R66)`)

### Styles and Shared Variables
* **`ComboBox` Styles**: Added a new SCSS file for styling the `ComboBox` component, including dropdown, input, and button styles. (`src/components/ComboBox/_combobox.scss`, `[src/components/ComboBox/_combobox.scssR1-R101](diffhunk://#diff-636cfe48edda0b84c22b986c75b813027a1d0095954c2250b0ce380a2cd50439R1-R101)`)
* **Shared Variables**: Introduced additional color variables (`$soft-gray`, `$pale-gray`) and label size variables (`--label-size4`) to support the new components. (`src/scss/_vars.scss`, `[[1]](diffhunk://#diff-bc8388c657acf6c610df95fb7fff1ed0191c845b9dd703d7d0ac198085b881b9R21-R22)`, `[[2]](diffhunk://#diff-bc8388c657acf6c610df95fb7fff1ed0191c845b9dd703d7d0ac198085b881b9R154)`, `[[3]](diffhunk://#diff-bc8388c657acf6c610df95fb7fff1ed0191c845b9dd703d7d0ac198085b881b9R199)`)

### Tests and Storybook Entries
* **Tests**: Added comprehensive unit tests for both `ComboBox` and `PhoneNumberPicker` components, covering various user interactions and edge cases. (`src/components/ComboBox/ComboBox.test.tsx`, `[[1]](diffhunk://#diff-f4c31db0a12a3dfae3b7f0b70a212d140b8953c67e006269caf1ee58b99f431aR1-R145)`; `src/components/PhoneNumberPicker/PhoneNumberPicker.test.tsx`, `[[2]](diffhunk://#diff-d028b2215265501ba1f98b0651db8f286472722aa643ec96fd1f4d38a3d97b20R1-R57)`)
* **Storybook**: Added Storybook entries for `ComboBox` and `PhoneNumberPicker` to demonstrate their usage and customization options. (`src/components/ComboBox/ComboBox.stories.tsx`, `[[1]](diffhunk://#diff-20109226114674d39e1a4f4114f7f1432958a60632fb813452d922a8695557b1R1-R33)`; `src/components/PhoneNumberPicker/PhoneNumberPicker.stories.tsx`, `[[2]](diffhunk://#diff-bc096ac0e0e73bc8c93ae796716f190c41d405ad9f843066b13c4ae4729eb84cR1-R25)`)

### Dependency Updates
* **Dependencies**: Added `cmdk` and `libphonenumber-js` libraries to support the new components. (`package.json`, `[package.jsonR48-R62](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R48-R62)`)

### Integration
* **Exports**: Updated the main export files to include the new components. (`src/index.ts`, `[[1]](diffhunk://#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R81)`; `src/components/ComboBox/index.ts`, `[[2]](diffhunk://#diff-9b7ff1a50dd6de12b918bf7a873f75b2771e916ff752dc06345c03d0e3e85a12R1)`)
* **SCSS Imports**: Imported the `ComboBox` styles into the global SCSS file. (`src/componentStyles.scss`, `[src/componentStyles.scssR41](diffhunk://#diff-7cd020c1b5b4444625c4d9819729ef3991b64ae9389d021fa334c44929fde889R41)`)

**Acceptance Test (how to verify the PR)**

- Add step by step instructions on how to verify the change

**Regression Test**

- (Optional) Add verification steps to make sure this PR doesn't break old functionality

**Evidence of testing**

- Post logs, screenshots, etc

**Things to look for during review**

- [ ] PR title should correctly describe the most significant type of commit. I.e. `feat(scope): ...` if a `minor` release should be triggered.
- [ ] All commit messages follow convention and are appropriate for the changes
- [ ] All references to `phillips` class prefix are using the prefix variable
- [ ] All major areas have a `data-testid` attribute.
- [ ] Document all props with jsdoc comments
- [ ] All strings should be translatable.
- [ ] Unit tests should be written and should have a coverage of 90% or higher in all areas.

New Components

- [ ] Are there any [accessibility considerations](https://www.w3.org/WAI/ARIA/apg/patterns/) that need to be taken into account and tested?
- [ ] Default story called "Playground" should be created for all new components
- [ ] Create a jsdoc comment that has an Overview section and a link to the Figma design for the component
- [ ] Export the component and its typescript type from the `index.ts` file
- [ ] Import the component scss file into the `componentStyles.scss` file.

[L3-6409]: https://phillipsauctions.atlassian.net/browse/L3-6409?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[L3-5961]: https://phillipsauctions.atlassian.net/browse/L3-5961?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/ComboBox/ComboBox.tsx [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+    document.addEventListener('mousedown', handleClickOutside);
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
+  }, [inputValue, sanitizedOptions, baseClassName, setInputValue]);
+
+  return (
+    <div
+      ref={(node) => {
```
Comment (scottdickerson) [inline src/components/ComboBox/ComboBox.tsx:76]: ah, are you trying to mergeRefs here?  There's a good utility function to do that `https://floating-ui.com/docs/react-utils`

File: src/components/ComboBox/ComboBox.tsx [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+            aria-label={`${id}-dropdown`}
+            className={`${baseClassName}__dropdown-button`}
+            onClick={() => setIsOpen((prev) => !prev)}
+            data-testid={`${id}-dropdown`}
+          >
+            <Icon color="$pure-black" height={18} icon="ChevronDown" width={18} className={`${baseClassName}__icon`} />
+          </button>
+          <Popover.Portal>
+            <Popover.Content className={`${baseClassName}__content`} aria-label={`${id}-content`}>
+              <CommandList className={`${baseClassName}__list`} hidden={!isOpen}>
```
Comment (scottdickerson) [inline src/components/ComboBox/ComboBox.tsx:139]: I'm not sure this is what we want, this keeps the command list in the dom even when it's not expanded.  can't we just have the PopoverRoot open change?

File: src/components/ComboBox/ComboBox.tsx [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+              aria-label={`${id}-input`}
+              data-testid={`${id}-input`}
+            />
+          </Popover.Trigger>
+          {inputValue.length > 0 && (
+            <button
+              className={`${baseClassName}__close-button`}
+              data-testid={`${id}-item-close-button`}
+              onClick={() => setInputValue('')}
+              aria-label={`${id}-clear`}
```
Comment (scottdickerson) [inline src/components/ComboBox/ComboBox.tsx:124]: these should be translated props

File: src/components/ComboBox/ComboBox.tsx (hunk 2 above)
Comment (scottdickerson) [reply src/components/ComboBox/ComboBox.tsx:139]:
```suggestion
      {isOpen ? (
                <CommandList className={`${baseClassName}__list`}>
```

File: src/components/ComboBox/ComboBox.tsx [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+   * Input value for the ComboBox.
+   */
+  inputValue: string;
+  /**
+   * Passed in function to handle input value changes.
+   */
+  setInputValue: (value: string) => void;
+}
+
+const ComboBox = React.forwardRef<HTMLDivElement, ComboBoxProps>(function ComboBox(
```
Comment (scottdickerson) [inline src/components/ComboBox/ComboBox.tsx:41]: can you add the link to the Figma here and storybook for this component

**Comment by scottdickerson** [comment]
accessibility is working well, with one small exception.  After I select a value then try and tab out of the field, it seems to focus on the "Clear" button with no focus indicator before leaving the field.  I think both the Clear button and the Dropdown button can be excluded from the tab index since interacting with the input field itself is the trigger we want.  After I select the value in this video I clicked the tab 3 times before it finally let me leave the field and focused on "Show code"

https://github.com/user-attachments/assets/ab9d74ca-6caf-4c0d-9c5d-f276e2399385

File: src/components/ComboBox/ComboBox.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+   */
+  id: string;
+  /**
+   * Optional className for custom styling.
+   */
+  className?: string;
+  /**
+   * Label for the ComboBox.
+   */
+  label: string;
```
Comment (davidicus) [inline src/components/ComboBox/ComboBox.tsx:26]: can we align with the other input prop names? I believe this is `labelText`. Both our input and select have a `hideLabel` prop that we should probably support here as well.

File: src/components/ComboBox/_combobox.scss [hunk 6]
```diff
@@ -0,0 +1,101 @@
+@use '#scss/allPartials' as *;
+
+.#{$px}-combo-box {
+  position: relative;
+  width: 100%;
+
+  &__input {
+    border: 1px solid $keyline-gray;
+    border-radius: 5px;
+    height: 40px;
```
Comment (davidicus) [inline src/components/ComboBox/_combobox.scss:10]: I would rather not have an explicit height if we can help it

**Review by scottdickerson — CHANGES_REQUESTED** [review]
A few more requested changes:  Hide the Clear button if the field is not focused or hovered, and fix the css positioning for the input buttons

**Comment by zhenpanda** [comment]
https://github.com/user-attachments/assets/20b453c2-d321-41a2-a81c-847e0f641790

File: src/components/PhoneNumberPicker/PhoneNumberPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+            return {
+              label: country,
+              value: `+${getCountryCode(country)}`,
+            };
+          })}
+          {...props}
+          inputValue={inputValue}
+          setInputValue={(value) => {
+            setInputValue(value.split(' ').pop() || '');
+          }}
```
[the remark this answers is not in the document]
Comment (zhenpanda) [reply src/components/PhoneNumberPicker/PhoneNumberPicker.tsx:65]: This is now address in the combo box component, phoneNumberPicker just has to pass in the correct { label, value }

https://github.com/user-attachments/assets/c96832cb-df4c-4feb-bcaf-4a98f064780d

File: src/components/ComboBox/ComboBox.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    <div style={{ height: '70px', width: '400px' }}>
+      <ComboBox {...props} inputValue={inputValue} setInputValue={setInputValue} />
+    </div>
+  );
+};
+
+Playground.args = {
+  options: birthdays,
+  id: 'birthdays-combo-box',
+  label: 'Birth Year',
```
[the remark this answers is not in the document]
Comment (zhenpanda) [reply src/components/ComboBox/ComboBox.stories.tsx]: Thanks for catching that, should be fixed now

File: src/components/ComboBox/ComboBox.tsx [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+      {...props}
+    >
+      <label htmlFor={`${id}-input`} className={`${baseClassName}__label`} data-testid={`${id}-label`}>
+        {label}
+      </label>
+      <Command
+        loop
+        shouldFilter={false}
+        onKeyDown={(e) => {
+          setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (sgrund14) [reply src/components/ComboBox/ComboBox.tsx:118]: might leave a comment in their repo about that, or see if there ever was an issue about the same thing 🤔

File: src/components/ComboBox/ComboBox.test.tsx [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    await userEvent.click(trigger);
+    const input = screen.getAllByTestId('test-id-input')[0];
+
+    await userEvent.click(input);
+    await userEvent.type(input, 'HI');
+
+    expect(screen.queryByText('No Options.')).toBeInTheDocument();
+  });
+
+  it('should clear input value when the close button is clicked', async () => {
```
Comment (davidicus) [inline src/components/ComboBox/ComboBox.test.tsx:84]: This has a similar test name to the test below and this test does not click on the close button
```suggestion
  it('should clear input value when user clicks outside of input', async () => {
```

It is debatable if this is desirable UX. What if the user wanted to copy text from somewhere else on the page and include? Can ask the team what they think. Not blocking approval over it.

File: src/components/ComboBox/ComboBox.tsx [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+  );
+
+  return (
+    <div ref={ref} className={classnames(baseClassName, className)} id={id} {...commonProps} {...props}>
+      <label htmlFor={`${id}-input`} className={`${baseClassName}__label`} data-testid={`${id}-label`}>
+        {labelText}
+      </label>
+      <Command
+        loop
+        shouldFilter={false}
```
Comment (sgrund14) [inline src/components/ComboBox/ComboBox.tsx:116]: why opt out of the built in filtering?

File: src/components/ComboBox/ComboBox.tsx [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+import { getCommonProps } from '../../utils';
+import { getScssVar } from '../../utils/scssUtils';
+
+export interface ComboBoxProps {
+  /**
+   * List of options to be displayed in the ComboBox.
+   */
+  options: {
+    label?: string;
+    value: string;
```
Comment (davidicus) [inline src/components/ComboBox/ComboBox.tsx:16]: We may want to revisit this in the future to be able to support passing in react nodes for use cases of text and icons.

File: src/components/ComboBox/ComboBox.tsx [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+   */
+  id: string;
+  /**
+   * Optional className for custom styling.
+   */
+  className?: string;
+  /**
+   * Label for the ComboBox.
+   */
+  labelText: string;
```
Comment (davidicus) [inline src/components/ComboBox/ComboBox.tsx:29]: We are still missing the hideLabelText prop to be consistent with other input components

File: src/components/ComboBox/ComboBox.tsx (hunk 11 above)
Comment (zhenpanda) [reply src/components/ComboBox/ComboBox.tsx:116]: We want it here to provides flexibility to filter options based on specific requirements, such as matching by both applying case-insensitive matching and matching both by label and value.

File: src/components/ComboBox/ComboBox.tsx (hunk 13 above)
Comment (zhenpanda) [reply src/components/ComboBox/ComboBox.tsx:29]: Yea @smarks8 mentioned, I will add it

File: src/components/ComboBox/ComboBox.tsx (hunk 12 above)
Comment (davidicus) [reply src/components/ComboBox/ComboBox.tsx:16]: Not a blocker

File: src/components/ComboBox/ComboBox.tsx (hunk 11 above)
Comment (sgrund14) [reply src/components/ComboBox/ComboBox.tsx:116]: ah got it

fwiw, there is a custom filter function prop detailed in the README docs https://github.com/pacocoursey/cmdk

```
You can provide a custom filter function that is called to rank each item. Note that the value will be trimmed.

<Command
  filter={(value, search) => {
    if (value.includes(search)) return 1
    return 0
  }}
/>
A third argument, keywords, can also be provided to the filter function. Keywords act as aliases for the item value, and can also affect the rank of the item. Keywords are trimmed.

<Command
  filter={(value, search, keywords) => {
    const extendValue = value + ' ' + keywords.join(' ')
    if (extendValue.includes(search)) return 1
    return 0
  }}
/>
```

but also it supports manual filtering so 🤷

File: src/components/ComboBox/ComboBox.tsx [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+  const CloseIcon = iconComponents['CloseX'];
+
+  const itemRefs = React.useRef<Record<string, HTMLDivElement | null>>({});
+  const filteredOptions = options.filter(
+    (option) =>
+      option.value.toLowerCase().includes(inputValue.toLowerCase()) ||
+      (option.label && option.label.toLowerCase().includes(inputValue.toLowerCase())),
+  );
+
+  useClickOutside(
```
Comment (constantinehuzenko) [inline src/components/ComboBox/ComboBox.tsx:105]: https://usehooks-ts.com/react-hook/use-on-click-outside

File: package.json [hunk 15]
```diff
@@ -42,22 +42,25 @@
   },
   "dependencies": {
     "@artsy/fresnel": "^7.1.4",
+    "@mantine/hooks": "^8.0.0",
```
[the remark this answers is not in the document]
Comment (zhenpanda) [reply package.json:45]: yea i can remove it if it's already in the code base

File: src/components/PhoneNumberPicker/PhoneNumberPicker.tsx [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+ * Overview of PhoneNumberPicker component
+ *
+ * [Figma Link] https://www.figma.com/design/rIefa3bRPyZbZmtyV9PSQv/My-Account?node-id=1-3&p=f&m=dev
+ *
+ * [Storybook Link](https://phillips-seldon.netlify.app/?path=/docs/components-PhoneNumberPicker--overview)
+ */
+const PhoneNumberPicker = React.forwardRef<HTMLDivElement, PhoneNumberPickerProps>(
+  ({ inputValue, labelText, className, id, setInputValue, ...props }, ref) => {
+    const { className: baseClassName, ...commonProps } = getCommonProps({ id }, 'PhoneNumberPicker');
+    const countries = useMemo(() => getCountries(), []);
```
Comment (constantinehuzenko) [inline src/components/PhoneNumberPicker/PhoneNumberPicker.tsx:46]: do we need useMemo here if we don't have dependencies?

**Comment by constantinehuzenko** [comment]
Screenshot 2025-05-08 at 1 17 00 PM
arrow's broken

**Comment by constantinehuzenko** [comment]
https://github.com/user-attachments/assets/f26a509f-1317-43c8-a082-967e765a0916

should it remove value when you close?

**Comment by zhenpanda** [comment]
@constantinehuzenko yes, this is how material UI handles a no option value in input while clicking dropdown close, when user clicks away the input will reset to blank if not a valid option.

**Comment by scottdickerson** [comment]
useClickOutside doesn't close the popover if a valid value has been typed


Uploading Screen Recording 2025-05-08 at 3.12.44 PM.mov…

**Comment by zhenpanda** [comment]
https://github.com/user-attachments/assets/b8315dda-895d-41bd-b167-5d03d4041810

**Review by scottdickerson — APPROVED** [review]
LGTM thanks for all the work on the accessibility
